cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -377,52 +349,51 @@ private void publishKillProgram(String stopPayload) {
   }
 
   /**
-   * Returns program status updates for tethered peers.
+   * Reads from per-peer topic and returns program status updates for the given tethered peer.
    */
-  private PeerProgramUpdates getPeerProgramUpdates(String afterMessageId) {
-    Map<String, List<Notification>> peerToNotifications = new HashMap<>();
-    String lastMessageId = afterMessageId;
+  private ProgramStateUpdates getProgramStateUpdates(String peerName) {
+    String topic = programStateTopicPrefix + peerName;
+    String lastMessageId = lastMessageIdForPeer.get(peerName);
+    List<Notification> notifications = new ArrayList<>();
     try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
-                                                                    programUpdateTopic,
+                                                                    topic,
                                                                     programUpdateFetchSize,
-                                                                    afterMessageId)) {
+                                                                    lastMessageIdForPeer.get(peerName))) {
       while (iterator.hasNext()) {
         Message message = iterator.next();
-        Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
-        if (notification.getNotificationType() == Notification.Type.PROGRAM_STATUS) {
-          Map<String, String> properties = notification.getProperties();
-          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
-          try {
-            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
-            if (runRecord.getPeerName() != null) {
-              String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
-              LOG.debug("Notifying peer {} about program run {} in state {}",
-                       runRecord.getPeerName(), programRunId, programStatus);
-              peerToNotifications.computeIfAbsent(runRecord.getPeerName(), n -> new ArrayList<>()).add(notification);
-            }
-          } catch (NotFoundException | IOException e) {
-            LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
-          }
-        }
+        notifications.add(message.decodePayload(r -> GSON.fromJson(r, Notification.class)));
         lastMessageId = message.getId();
       }
-    } catch (Exception e) {
-      LOG.error("Exception when fetching program updates. Will retry again during next poll", e);
+    } catch (TopicNotFoundException e) {
+      createTopicIfNeeded(new TopicId(NamespaceId.SYSTEM.getNamespace(), topic));
```

Comment:
the subscriber shouldn't be creating the topic since it never writes to it. Whatever publishes the messages is the one that should react to TopicNotFound by creating the topic.

Reply:
removed topic creation logic here